lib/puppet/util/command_line/trollop.rb

```diff
@@ -335,9 +335,21 @@ def parse cmdline = ARGV
               when /^-([^-])$/
                 @short[::Regexp.last_match(1)]
               when /^--no-([^-]\S*)$/
-                @long["[no-]#{::Regexp.last_match(1)}"]
+                possible_match = @long["[no-]#{::Regexp.last_match(1)}"]
+                if !possible_match
+                  Puppet.warning _("Partial argument match detected: %{arg}. Partial argument matching will be deprecated in Puppet 9.") % { arg: arg }
+                  @long["[no-]#{::Regexp.last_match(1).tr('-', '_')}"] || @long["[no-]#{::Regexp.last_match(1).tr('_', '-')}"]
```

Comment:
Sorry, I left a comment, but it was marked as "pending" so maybe it wasn't visible? In any case, I think there are a couple adjustments needed:

* The warning is triggered anytime an application-specific option is given: 

    ```
    $ bundle exec puppet apply --execute "notice(0)"
    Warning: Partial argument match detected: --execute. Partial argument matching will be deprecated in Puppet 9.
    ```
    I think we only want to trigger the warning if an argument is given that doesn't match a global option, but it does match one of the translated versions. So `--execute` is not a global option (it's specific to the `apply` application) and shouldn't trigger the warning. The argument `--show-diff` is also not a global option, but it does match the translated version `--show_diff`. So in that case we do want to print the warning.

* If the warning is triggered, it would be good to specify what option the user *should* be using, so they can self-correct. For example, if a class expects an argument of type `String`, but you pass it an `Integer`, we print both the expected and given types:

    ```
    Class[One]: parameter 'a' expects a String value, got Integer
    ```

* Could you use `Puppet.deprecation_warning` instead of `Puppet.warning` so that the deprecation (along with all other deprecations) can be suppressed using `puppet apply --disable_warnings deprecations ...`?

* Technically, partial argument matches are deprecated now (as of this PR) and the functionality will be removed in a future release. For example, this is the message we print for deprecated puppet applications: https://github.com/puppetlabs/puppet/blob/e9bccaeb017096836214dbb983c47ab4dc96746d/lib/puppet/application/face_base.rb#L255